File: snake-game/csharp/src/SnakeGame/Game.cs
namespace SnakeGame;

/// <summary>
/// Aggregate: board + snake + food + score + state.
/// <c>Tick()</c> advances the game by one step.
/// </summary>
public class Game
{
    private readonly int _width;
    private readonly int _height;
    private readonly Func<IReadOnlyList<Position>, Position> _foodSpawner;

    public Game(int width, int height, Func<IReadOnlyList<Position>, Position> foodSpawner)
    {
        _width = width;
        _height = height;
        _foodSpawner = foodSpawner;
        Snake = new Snake(new Position(0, 0), Direction.Right);
        Score = 0;
        State = GameState.Playing;
        Food = foodSpawner(EmptyCells());
    }

    internal Game(int width, int height,
                  Func<IReadOnlyList<Position>, Position> foodSpawner,
                  Snake snake, Position food, int score, GameState state)
    {
        _width = width;
        _height = height;
        _foodSpawner = foodSpawner;
        Snake = snake;
        Food = food;
        Score = score;
        State = state;
    }

    public Snake Snake { get; private set; }
    public Position Food { get; private set; }
    public int Score { get; private set; }
    public GameState State { get; private set; }

    public void ChangeDirection(Direction newDirection)
    {
        if (State != GameState.Playing) return;
        Snake = Snake.ChangeDirection(newDirection);
    }

    public void Tick()
    {
        if (State != GameState.Playing) return;

        var newHead = Snake.Direction.Move(Snake.Head);

        if (IsOutOfBounds(newHead) || IsBodyCollision(newHead))
        {
            State = GameState.GameOver;
            return;
        }

        var eatsFood = newHead == Food;
        Snake = Snake.Move(eatsFood);

        if (eatsFood)
        {
            Score++;

            if (Snake.Length == _width * _height)
            {
                State = GameState.Won;
                return;
            }

            Food = _foodSpawner(EmptyCells());
        }
    }

    private bool IsOutOfBounds(Position position) =>
        position.X < 0 || position.X >= _width ||
        position.Y < 0 || position.Y >= _height;

    private bool IsBodyCollision(Position position) =>
        Snake.OccupiesPosition(position);

    private IReadOnlyList<Position> EmptyCells()
    {
        var empty = new List<Position>();
        for (var x = 0; x < _width; x++)
        for (var y = 0; y < _height; y++)
        {
            var pos = new Position(x, y);
            if (!Snake.OccupiesPosition(pos))
                empty.Add(pos);
        }
        return empty;
    }
}

File: snake-game/typescript/src/Game.ts
import type { Position } from './Position.js';
import { positionsEqual } from './Position.js';
import type { Direction } from './Direction.js';
import { move } from './Direction.js';
import { Snake } from './Snake.js';

export type GameState = 'Playing' | 'GameOver' | 'Won';
export type FoodSpawner = (emptyCells: Position[]) => Position;

/**
 * Aggregate: board + snake + food + score + state.
 * `tick()` advances the game by one step.
 */
export class Game {
  private readonly width: number;
  private readonly height: number;
  private readonly foodSpawner: FoodSpawner;

  snake: Snake;
  food: Position;
  score: number;
  state: GameState;

  constructor(
    width: number,
    height: number,
    foodSpawner: FoodSpawner,
    snake?: Snake,
    food?: Position,
    score?: number,
    state?: GameState,
  ) {
    this.width = width;
    this.height = height;
    this.foodSpawner = foodSpawner;
    this.snake = snake ?? new Snake([{ x: 0, y: 0 }], 'Right');
    this.score = score ?? 0;
    this.state = state ?? 'Playing';
    this.food = food ?? foodSpawner(this.emptyCells());
  }

  changeDirection(newDirection: Direction): void {
    if (this.state !== 'Playing') return;
    this.snake = this.snake.changeDirection(newDirection);
  }

  tick(): void {
    if (this.state !== 'Playing') return;

    const newHead = move(this.snake.direction, this.snake.head);

    if (this.isOutOfBounds(newHead) || this.isBodyCollision(newHead)) {
      this.state = 'GameOver';
      return;
    }

    const eatsFood = positionsEqual(newHead, this.food);
    this.snake = this.snake.move(eatsFood);

    if (eatsFood) {
      this.score++;

      if (this.snake.length === this.width * this.height) {
        this.state = 'Won';
        return;
      }

      this.food = this.foodSpawner(this.emptyCells());
    }
  }

  private isOutOfBounds(position: Position): boolean {
    return (
      position.x < 0 ||
      position.x >= this.width ||
      position.y < 0 ||
      position.y >= this.height
    );
  }

  private isBodyCollision(position: Position): boolean {
    return this.snake.occupiesPosition(position);
  }

  private emptyCells(): Position[] {
    const empty: Position[] = [];
    for (let x = 0; x < this.width; x++) {
      for (let y = 0; y < this.height; y++) {
        const pos: Position = { x, y };
        if (!this.snake.occupiesPosition(pos)) {
          empty.push(pos);
        }
      }
    }
    return empty;
  }
}
